server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/analysis/AsyncAnalysisService.java

```diff
@@ -114,47 +116,71 @@ public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(Strin
    */
-  public synchronized CompletableFuture<CobolDocumentModel> scheduleAnalysis(String uri, String text, Integer currentRevision, boolean open, boolean force, SourceUnitGraph.EventSource eventSource) {
+  public synchronized FutureTask<CobolDocumentModel> scheduleAnalysis(String uri, String text, Integer currentRevision, boolean open, boolean force, SourceUnitGraph.EventSource eventSource) {
     notifyAllListeners(AnalysisState.SCHEDULED, documentModelService.get(uri), eventSource);
     String id = makeId(uri, currentRevision);
     Integer prevId = analysisResultsRevisions.put(uri, currentRevision);
     if (currentRevision.equals(prevId) && !force) {
       notifyAllListeners(AnalysisState.SKIPPED, documentModelService.get(uri), eventSource);
       return analysisResults.get(id);
     }
-    Executor analysisExecutor = getExecutor(uri);
-    CompletableFuture<CobolDocumentModel> value = CompletableFuture.supplyAsync(() -> {
+    ExecutorService analysisExecutor = getExecutor(uri);
+    CobolDocumentModel documentModel = documentModelService.get(uri);
+    if (documentModel.getLastAnalysisResult() != null) {
+      // initiate cancel on prev running analysis
+      cancelRunningAnalysis(ImmutableList.of(documentModel));
+    }
+    FutureTask<CobolDocumentModel> futureTask = new FutureTask<>(scheduleAnalysis(uri, documentModel, text, currentRevision, open, force, eventSource, id));
+    analysisResults.put(id, futureTask);
+    analysisExecutor.submit(futureTask);
+    if (prevId != null && !force) {
+      Optional.ofNullable(analysisResults.get(makeId(uri, prevId))).ifPresent(cf -> cf.cancel(true));
+    }
+    return futureTask;
+  }
+
+  private Callable<CobolDocumentModel> scheduleAnalysis(String uri, CobolDocumentModel documentModel, String text, Integer currentRevision, boolean open, boolean force, SourceUnitGraph.EventSource eventSource, String id) {
+    return () -> {
       if (currentRevision < analysisResultsRevisions.get(uri) && !force) {
-        notifyAllListeners(AnalysisState.SKIPPED, documentModelService.get(uri), eventSource);
-        LOG.debug("[scheduleAnalysis] skip revision: " + currentRevision + " latest: " + analysisResultsRevisions.get(uri));
+        notifyAllListeners(
+                AnalysisState.SKIPPED, documentModelService.get(uri), eventSource);
+        LOG.debug(
+                "[scheduleAnalysis] skip revision: "
+                        + currentRevision
+                        + " latest: "
+                        + analysisResultsRevisions.get(uri));
         return null;
       }
-      LOG.debug("[scheduleAnalysis] Start analysis: " + uri);
-      notifyAllListeners(AnalysisState.STARTED, documentModelService.get(uri), eventSource);
+      LOG.debug("[scheduleAnalysis] waiting for previous analysis of {} to finish", uri);
       try {
+        if (documentModel.getLastAnalysisResult() != null) {
+          waitForPreviousAnalysisToBeCompleted(ImmutableList.of(uri));
+        }
```

Comment:
Should the waiting be unconditional? The rule is that we don't want to cancel the first parsing, however we still don't want run two parsings side by side, right?